src/main/daml/Daml/Finance/Instrument/StructuredProduct/Util.daml

```diff
@@ -0,0 +1,75 @@
+
+type O = Observation Date Decimal Observable
+type C = Claim Date Decimal Deliverable Observable
+
+-- | Find out which schedule periods of scheduleA exist in scheduleB.
+includes : Schedule -> Schedule -> [Bool]
+includes scheduleA scheduleB =
+  map (\a -> isSome $ find (\b -> b.adjustedEndDate == a.adjustedEndDate) scheduleB) scheduleA
+
+-- | Calculate the claims for an auto-callable with a contingent coupon on each payment date
+--   and a redemption amount at maturity (unless auto-called previously).
+createAutoCallableClaims : (Date -> Time) -> Schedule -> Schedule -> PeriodicSchedule
+  -> Bool -> Decimal -> DayCountConventionEnum -> Decimal -> Deliverable -> Optional FloatingRate
+  -> Optional Decimal -> Optional Decimal -> Text -> Decimal -> Decimal -> Decimal -> Decimal
+  -> HolidayCalendarData -> TaggedClaim
+createAutoCallableClaims dateToTime paymentSchedule callableSchedule periodicSchedule
+  useAdjustedDatesForDcf couponRate dayCountConvention notional cashInstrument floatingRate
+  capRate floorRate spot couponBarrier callBarrier finalBarrier putStrike
+  fixingCalendars =
+  let
+    notionalAmount = scale (Const notional) $ one cashInstrument
+
+    principal = notionalAmount
+    initialObservationDate = (.adjustedStartDate) $ head callableSchedule
+
+    combineTagClaim (couponPeriod, callPeriod) notCalledClaim =
+      let
+        cpn = calculateRatePayment couponPeriod dayCountConvention useAdjustedDatesForDcf
+          periodicSchedule floatingRate couponRate notionalAmount fixingCalendars capRate floorRate
+        (callDate, paymentDate) = if callPeriod.adjustedEndDate > couponPeriod.adjustedEndDate
+          then error "each call date must be before or at the corresponding payment date"
+          else (callPeriod.adjustedEndDate, couponPeriod.adjustedEndDate)
+        spotOnObservationDate = ObserveAt spot callDate
+        couponBarrierLevel = Const couponBarrier * ObserveAt spot initialObservationDate
+        couponBarrierHit = Lte (spotOnObservationDate, couponBarrierLevel)
+        coupon = cond couponBarrierHit zero cpn
+        called = when (at paymentDate) $ and coupon principal
+        notCalled = when (at paymentDate) $ and coupon notCalledClaim
+        callBarrierLevel = Const callBarrier * ObserveAt spot initialObservationDate
+        autoExerciseCondition = Lte (callBarrierLevel, spotOnObservationDate)
+        tailClaim = when (at callDate) $ cond autoExerciseCondition called notCalled
+      in
+        tailClaim
+
+    finalCouponDate = (.adjustedEndDate) $ last paymentSchedule
+    finalObservationDate = (.adjustedEndDate) $ last callableSchedule
+    spotOnObservationDate = ObserveAt spot finalObservationDate
+    finalBarrierLevel = Const finalBarrier * ObserveAt spot initialObservationDate
+    barrierHit = Lte (spotOnObservationDate, finalBarrierLevel)
+    putStrikeLevel = Const putStrike * ObserveAt spot initialObservationDate
+    perf = spotOnObservationDate / putStrikeLevel
```

Comment:
not sure this is right. 

performance = spotOnObservationDate / ObserveAt spot initialObservationDate

If the barrier is hit, the final payment should be be [ 1 - max(strike - perf, 0) / strike] = [ 1 - max(1 - perf / strike, 0)]

Reply:
In order to capture the error, we would need a test scenario with a strike other than 100%

Reply:
I wrote this with the sample termsheet in mind, where the strike matches the initial fixing and is not really given as a variable. Seems that we are deviating from that now, by changing the absolute initialFixing to a percentage and then renaming it to a putStrike. We may need that in the future, but that feature was never intended. 

I have added a check for now that the strike is indeed 100%. Will add an issue to possibly extend the functionality as well, as part of a separate PR.

Reply:
https://github.com/digital-asset/daml-finance/issues/1196

Reply:
By the way, I also opened this issue earlier: https://github.com/digital-asset/daml-finance/issues/1194
We may need that when the observation dates do not match a regular schedule (which seems to be the case for the sample product).

Reply:
Thanks! I would suggest to add a comment to the formula pointing to the issue and mentioning that it doesn't work when strike is other than 100%

> By the way, I also opened this issue earlier: #1194 We may need that when the observation dates do not match a regular schedule (which seems to be the case for the sample product).

Yes, I noticed that it was using custom dates. Does this mean that we currently can't reproduce the exact trade because those dates are not originated from a regular schedule?

Reply:
It probably depends on the holiday calendars used, but for a typical NYSE holiday calendar and a regular schedule I think there is at least one day that would be off.